File: sdks/python/src/fireweave/adapters/posthog.py
"""PostHog backend adapter (requires the ``fireweave[posthog]`` extra).

Wraps ``posthog==7.31.0`` per ADR-0002:

- **Snapshot evaluation**: uses ``Client.evaluate_flags()`` — one snapshot per
  evaluation context; per-flag reads never trigger extra network calls or
  vendor-side ``$feature_flag_called`` events (Fireweave owns exposures).
- **Modes**: remote (``phc_`` project key -> /flags?v=2) or local evaluation
  (``phs_`` secret key / ``phx_`` personal key -> definitions poll).
- **Owned vs injected client**: pass ``client=`` to inject a preconfigured
  ``posthog.Posthog`` (the adapter will NOT shut it down), or pass config and
  the adapter owns the client lifecycle.
- **No vendor types in the public API**: this module returns only
  :class:`~fireweave.adapters.base.FlagResolution` and Fireweave errors.
- **Transport injection**: ``transport=`` overrides the snapshot fetcher for
  fault testing; transport exceptions are mapped to the canonical taxonomy by
  :func:`map_transport_error`.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, Optional, Protocol

from ..config import FireweaveConfig
from ..context import EvaluationContext
from ..errors import (
    AuthenticationError,
    AuthorizationError,
    BackendUnavailableError,
    ConfigurationError,
    FireweaveError,
    FlagNotFoundError,
    MalformedResponseError,
    NetworkError,
    NotReadyError,
    RateLimitedError,
    TimeoutError_,
)
from .base import FlagResolution

__all__ = [
    "PostHogAdapter",
    "SnapshotTransport",
    "SnapshotData",
    "VendorFlagRecord",
    "map_transport_error",
]


@dataclass(frozen=True)
class VendorFlagRecord:
    """Vendor-neutral copy of one evaluated flag from a snapshot."""

    key: str
    enabled: bool
    variant: Optional[str] = None
    payload: Optional[Any] = None
    flag_id: Optional[int] = None
    version: Optional[int] = None
    reason: Optional[str] = None
    locally_evaluated: bool = False


@dataclass(frozen=True)
class SnapshotData:
    """Result of one snapshot fetch."""

    flags: Dict[str, VendorFlagRecord] = field(default_factory=dict)
    quota_limited: bool = False
    errors_while_computing: bool = False
    from_cache: bool = False


class SnapshotTransport(Protocol):
    """Fetches a flag snapshot for one identity; raises on transport failure."""

    def fetch(
        self,
        distinct_id: str,
        groups: Dict[str, str],
        person_properties: Dict[str, Any],
        group_properties: Dict[str, Dict[str, Any]],
    ) -> SnapshotData: ...


def map_transport_error(exc: Exception) -> FireweaveError:
    """Map transport exceptions to the canonical 15-kind taxonomy.

    The original exception is preserved on ``__cause__``; messages are the
    canonical safe defaults (no URLs, keys, or payload echoes).
    """
    mapped: FireweaveError
    status = _http_status(exc)
    if status is not None:
        if status == 401:
            mapped = AuthenticationError()
        elif status == 403:
            mapped = AuthorizationError()
        elif status == 429:
            mapped = RateLimitedError()
        elif status >= 500:
            mapped = BackendUnavailableError()
        else:
            mapped = BackendUnavailableError()
    elif isinstance(exc, (json.JSONDecodeError,)):
        mapped = MalformedResponseError()
    else:
        mapped = _map_requests_error(exc)
    mapped.__cause__ = exc
    return mapped


def _http_status(exc: Exception) -> Optional[int]:
    response = getattr(exc, "response", None)
    status = getattr(response, "status_code", None)
    if isinstance(status, int):
        return status
    status = getattr(exc, "status_code", None) or getattr(exc, "status", None)
    return status if isinstance(status, int) else None


def _map_requests_error(exc: Exception) -> FireweaveError:
    try:
        import requests.exceptions as rex
    except ImportError:  # pragma: no cover - requests ships with posthog
        return NetworkError()
    if isinstance(exc, rex.Timeout):
        return TimeoutError_()
    if isinstance(exc, rex.ConnectionError):
        return NetworkError()
    if isinstance(exc, rex.HTTPError):
        return BackendUnavailableError()
    if isinstance(exc, rex.RequestException):
        return NetworkError()
    if isinstance(exc, ValueError):
        return MalformedResponseError()
    return NetworkError()


class _ClientSnapshotTransport:
    """Default transport: ``posthog.Posthog.evaluate_flags`` snapshots.

    Reads the snapshot's internal records directly (no ``is_enabled`` /
    ``get_flag`` calls) so no vendor ``$feature_flag_called`` events fire —
    exposure telemetry is Fireweave-owned and deduped in the client.
    """

    def __init__(self, client: Any, *, only_evaluate_locally: bool) -> None:
        self._client = client
        self._only_locally = only_evaluate_locally

    def fetch(
        self,
        distinct_id: str,
        groups: Dict[str, str],
        person_properties: Dict[str, Any],
        group_properties: Dict[str, Dict[str, Any]],
    ) -> SnapshotData:
        snapshot = self._client.evaluate_flags(
            distinct_id,
            groups=groups or None,
            person_properties=person_properties or None,
            group_properties=group_properties or None,
            only_evaluate_locally=self._only_locally,
        )
        records: Dict[str, VendorFlagRecord] = {}
        for key, rec in getattr(snapshot, "_flags", {}).items():
            records[key] = VendorFlagRecord(
                key=key,
                enabled=bool(rec.enabled),
                variant=rec.variant,
                payload=_parse_payload(rec.payload),
                flag_id=rec.id,
                version=rec.version,
                reason=rec.reason,
                locally_evaluated=bool(rec.locally_evaluated),
            )
        return SnapshotData(
            flags=records,
            quota_limited=bool(getattr(snapshot, "_quota_limited", False)),
            errors_while_computing=bool(
                getattr(snapshot, "_errors_while_computing", False)
            ),
        )


def _parse_payload(payload: Any) -> Any:
    """PostHog payloads may arrive JSON-encoded; decode strings best-effort."""
    if isinstance(payload, str):
        try:
            return json.loads(payload)
        except (ValueError, TypeError):
            return payload
    return payload


# PostHog condition-match reason strings that indicate a targeting match.
_MATCH_REASONS = frozenset({"condition_match", "matched_condition", "super_condition_value"})


class PostHogAdapter:
    """BackendAdapter over PostHog remote (`phc_`) or local-eval (`phs_`/`phx_`)."""

    def __init__(
        self,
        *,
        client: Optional[Any] = None,
        config: Optional[FireweaveConfig] = None,
        transport: Optional[SnapshotTransport] = None,
    ) -> None:
        if client is None and config is None and transport is None:
            raise ConfigurationError(
                "PostHogAdapter requires a client, a config, or a transport",
                init_fatal=True,
            )
        self._config = config or FireweaveConfig()
        self._injected_client = client
        self._client: Optional[Any] = client
        self._owns_client = client is None and transport is None
        self._transport = transport
        self._initialized = False

    # -- lifecycle -----------------------------------------------------------

    def initialize(self) -> None:
        if self._transport is not None:
            self._initialized = True
            return
        if self._client is None:
            self._config.validate(backend_required=True)
            self._client = self._build_client()
        self._transport = _ClientSnapshotTransport(
            self._client,
            only_evaluate_locally=self._config.only_evaluate_locally,
        )
        self._initialized = True

    def _build_client(self) -> Any:
        try:
            from posthog import Posthog
        except ImportError as exc:
            raise ConfigurationError(
                "posthog extra not installed (pip install 'fireweave[posthog]')",
                init_fatal=True,
            ) from exc
        cfg = self._config
        kwargs: Dict[str, Any] = {
            "host": cfg.host,
            "feature_flags_request_timeout_seconds": (
                cfg.feature_flags_request_timeout_ms / 1000.0
            ),
            "enable_local_evaluation": cfg.local_evaluation,
            "sync_mode": False,
        }
        if cfg.personal_api_key:
            kwargs["personal_api_key"] = cfg.personal_api_key  # phx_
        if cfg.secret_key:
            kwargs["secret_key"] = cfg.secret_key  # phs_
        return Posthog(cfg.project_api_key, **kwargs)

    def shutdown(self, timeout_ms: int) -> None:
        self._initialized = False
        self._transport = None
        client, self._client = self._client, None
        if client is not None and self._owns_client:
            for method in ("flush", "shutdown"):
                fn = getattr(client, method, None)
                if callable(fn):
                    try:
                        fn()
                    except Exception:
                        pass

    # -- resolution ------------------------------------------------------------

    def resolve(self, flag_key: str, context: EvaluationContext) -> FlagResolution:
        if not self._initialized or self._transport is None:
            raise NotReadyError()

        distinct_id = context.targeting_key or ""
        plain = context.plain_attributes
        groups = context.groups
        group_properties = context.group_properties
        person_properties = {
            k: v
            for k, v in plain.items()
            if k not in ("groups", "groupProperties")
            and not k.startswith("fireweave.")
        }

        try:
            snapshot = self._transport.fetch(
                distinct_id, groups, person_properties, group_properties
            )
        except FireweaveError:
            raise
        except Exception as exc:
            raise map_transport_error(exc) from exc

        record = snapshot.flags.get(flag_key)
        if record is None:
            raise FlagNotFoundError(quota_limited=snapshot.quota_limited)

        # Value model: structured payload wins; else variant string; else the
        # boolean gate. Type conformance is enforced by the runtime.
        if record.payload is not None:
            value: Any = record.payload
        elif record.variant is not None:
            value = record.variant
        else:
            value = record.enabled

        return FlagResolution(
            value=value,
            variant=record.variant,
            enabled=record.enabled,
            matched=True,
            version=record.version,
            vendor_flag_id=record.flag_id,
            reason_code=record.reason,
            payload=record.payload,
            from_cache=snapshot.from_cache,
            quota_limited=snapshot.quota_limited,
        )

    # -- exposures ---------------------------------------------------------------

    def send_exposures(self, events: list) -> None:
        """Emit deduped Fireweave exposure events via the vendor client."""
        client = self._client
        if client is None:
            return
        capture = getattr(client, "capture", None)
        if not callable(capture):
            return
        for event in events:
            try:
                capture(
                    distinct_id=event.get("targetingKey"),
                    event="$feature_flag_called",
                    properties={
                        "$feature_flag": event.get("flagKey"),
                        "$feature_flag_response": event.get("value"),
                        "$feature_flag_variant": event.get("variant"),
                        "fireweave.rolloutId": event.get("rolloutId"),
                    },
                )
            except Exception:
                # Telemetry loss must never affect callers.
                pass

File: sdks/node/packages/sdk/src/adapters/posthog.ts
/**
 * PostHog backend adapter (ADR-0002). Wraps posthog-node@5.46.x behind the
 * canonical BackendAdapter boundary:
 *  - evaluateFlags() snapshot API only (no deprecated per-flag calls);
 *  - remote mode with a phc_ project key; local-eval / hybrid mode when a
 *    secret (phs_/phx_) key is supplied;
 *  - injected-vs-owned client lifecycle (injected clients are never shut down);
 *  - exposure policy: side-effect-free reads — local snapshot values are taken
 *    from internal `_flags` records (never `getFlag`/`isEnabled`, which emit
 *    vendor `$feature_flag_called`); owned clients also gate capture so stray
 *    vendor exposures without `fireweave.exposure` are dropped. Fireweave
 *    exposures flow only through {@link PostHogAdapter.recordExposure};
 *  - quotaLimited ⇒ FlagNotFound default + fireweave.quotaLimited metadata;
 *  - error mapping via an injected fetch observer, because posthog-node
 *    swallows /flags transport errors and returns empty snapshots.
 *  - hybrid local serve (secret key, no /flags observation) returns a Decision
 *    from the local snapshot — never maps success to Network (RB-1).
 *
 * No posthog-node types appear in this module's exported (public) API surface —
 * the client is accepted via the structural {@link PostHogClientLike} interface.
 */
import { FireweaveError } from '../errors.js';
import { assertHostAllowed } from '../hosts.js';
import type {
  AdapterResolution,
  AdapterRuntimeFeatures,
  BackendAdapter,
  ResolveOptions,
} from '../adapter.js';
import type { CanonicalContext, Exposure, JsonValue, Signal } from '../types.js';

/** Default bound on client shutdown (mirrors DEFAULT_SHUTDOWN_TIMEOUT_MS). */
const DEFAULT_ADAPTER_SHUTDOWN_TIMEOUT_MS = 10000;

/** Internal per-flag record shape (mirrors posthog-node EvaluatedFlagRecord). */
export interface SnapshotFlagRecord {
  key: string;
  enabled: boolean;
  variant?: string | null | undefined;
  payload?: unknown;
  id?: number | null | undefined;
  version?: number | null | undefined;
  reason?: string | null | undefined;
  locallyEvaluated?: boolean | undefined;
}

/** Snapshot returned by evaluateFlags — prefer `_flags` for side-effect-free reads. */
export interface EvaluateFlagsSnapshot {
  isEnabled(key: string): boolean;
  getFlag(key: string): string | boolean | undefined;
  getFlagPayload(key: string): unknown;
  keys: string[];
  /**
   * Internal flag records. Reading these does not emit `$feature_flag_called`
   * (unlike `getFlag` / `isEnabled`). Real posthog-node snapshots expose this;
   * test fakes should too when exercising the local path.
   */
  _flags?: Record<string, SnapshotFlagRecord>;
}

/** Minimal structural view of a posthog-node client (no vendor types leaked). */
export interface PostHogClientLike {
  evaluateFlags(distinctId: string, options?: {
    groups?: Record<string, string>;
    personProperties?: Record<string, string>;
    groupProperties?: Record<string, Record<string, string>>;
    onlyEvaluateLocally?: boolean;
    disableGeoip?: boolean;
    flagKeys?: string[];
  }): Promise<EvaluateFlagsSnapshot>;
  capture(props: {
    distinctId: string;
    event: string;
    properties?: Record<string, unknown>;
  }): void;
  waitForLocalEvaluationReady?(timeoutMs?: number): Promise<boolean>;
  isLocalEvaluationReady?(): boolean;
  flush(): Promise<void>;
  shutdown(shutdownTimeoutMs?: number): Promise<void>;
}

type FetchLike = (url: string, init?: {
  method?: string;
  headers?: Record<string, string>;
  body?: string | Uint8Array;
  signal?: AbortSignal;
}) => Promise<{
  status: number;
  text(): Promise<string>;
  json(): Promise<unknown>;
  headers: { get(name: string): string | null };
}>;

export interface PostHogAdapterOptions {
  /** PostHog project API key (phc_...). Required unless a client is injected. */
  projectApiKey?: string;
  /** PostHog host (e.g. the test-server stub or us.i.posthog.com). */
  host?: string;
  /**
   * SSRF allowlist override. Empty/undefined ⇒ the canonical PostHog default
   * allowlist + loopback; ['*'] opts out (self-hosted at your own risk).
   */
  allowedHosts?: readonly string[];
  /** Deadline handed to the vendor client's shutdown (default 10 000 ms). */
  shutdownTimeoutMs?: number;
  /** Secret/personal key (phs_/phx_) enabling local evaluation. */
  secretApiKey?: string;
  /** Evaluate exclusively from local definitions (no /flags fallback). */
  onlyEvaluateLocally?: boolean;
  /** Timeout for /flags requests. */
  featureFlagsRequestTimeoutMs?: number;
  /** Definitions polling interval (local eval). */
  featureFlagsPollingInterval?: number;
  /** Injected client (tests / shared client). Lifecycle stays with the caller. */
  client?: PostHogClientLike;
  /** Base fetch used under the observer (tests may stub the network here). */
  fetch?: FetchLike;
  /** Wait for the first definitions poll during initialize() (local eval). */
  waitForLocalDefinitions?: boolean;
}

interface FlagsObservation {
  kind: 'ok' | 'http' | 'network' | 'abort' | 'parse';
  status?: number;
  body?: FlagsV2Body;
}

interface FlagsV2Flag {
  key: string;
  enabled: boolean;
  variant: string | null;
  reason?: { code?: string; condition_index?: number | null; description?: string } | null;
  metadata?: { id?: number | null; version?: number | null; payload?: string | null } | null;
}

interface FlagsV2Body {
  flags?: Record<string, FlagsV2Flag>;
  errorsWhileComputingFlags?: boolean;
  requestId?: string;
  quotaLimited?: string[] | null;
}

function mapHttpStatus(status: number): FireweaveError {
  if (status === 401) return new FireweaveError('Authentication');
  if (status === 403) return new FireweaveError('Authorization');
  if (status === 429) return new FireweaveError('RateLimited');
  if (status >= 500) return new FireweaveError('BackendUnavailable');
  return new FireweaveError('BackendUnavailable');
}

export class PostHogAdapter implements BackendAdapter {
  readonly name = 'posthog' as const;
  private readonly options: PostHogAdapterOptions;
  private client: PostHogClientLike | undefined;
  private readonly ownsClient: boolean;
  private closed = false;
  /** Last /flags observation per distinct_id. */
  private readonly flagsObservations = new Map<string, FlagsObservation>();
  /** Local-eval definitions state (driven by the fetch observer). */
  private definitionsLoaded = false;
  private definitionsStale = false;

  constructor(options: PostHogAdapterOptions = {}) {
    this.options = { ...options };
    this.ownsClient = options.client === undefined;
    if (options.client !== undefined) {
      this.client = options.client;
    }
  }

  /** The fetch implementation handed to posthog-node: observes every request. */
  private buildObservedFetch(): FetchLike {
    const base: FetchLike = this.options.fetch ?? (globalThis.fetch as unknown as FetchLike);
    return async (url, init) => {
      const isFlags = url.includes('/flags') && !url.includes('/flags/definitions');
      const isDefinitions = url.includes('/flags/definitions');
      let distinctId: string | undefined;
      if (isFlags && typeof init?.body === 'string') {
        try {
          const parsed = JSON.parse(init.body) as { distinct_id?: string };
          distinctId = parsed.distinct_id;
        } catch {
          // ignore unparseable request bodies
        }
      }
      let response: Awaited<ReturnType<FetchLike>>;
      try {
        response = await base(url, init);
      } catch (err) {
        if (isFlags && distinctId !== undefined) {
          const aborted = err instanceof Error && err.name === 'AbortError';
          this.flagsObservations.set(distinctId, { kind: aborted ? 'abort' : 'network' });
        }
        if (isDefinitions && this.definitionsLoaded) this.definitionsStale = true;
        throw err;
      }
      if (isDefinitions) {
        if (response.status === 200) {
          this.definitionsLoaded = true;
          this.definitionsStale = false;
        } else if (response.status !== 304 && this.definitionsLoaded) {
          this.definitionsStale = true;
        }
        return response;
      }
      if (!isFlags || distinctId === undefined) return response;

      // Buffer the body so both we and posthog-node can read it.
      let text: string;
      try {
        text = await response.text();
      } catch {
        this.flagsObservations.set(distinctId, { kind: 'network' });
        throw new Error('response body read failed');
      }
      if (response.status >= 400) {
        this.flagsObservations.set(distinctId, { kind: 'http', status: response.status });
      } else {
        try {
          const body = JSON.parse(text) as FlagsV2Body;
          this.flagsObservations.set(distinctId, { kind: 'ok', status: response.status, body });
        } catch {
          this.flagsObservations.set(distinctId, { kind: 'parse', status: response.status });
        }
      }
      const buffered = {
        status: response.status,
        headers: response.headers,
        text: async () => text,
        json: async () => JSON.parse(text) as unknown,
      };
      return buffered as Awaited<ReturnType<FetchLike>>;
    };
  }

  async initialize(_signal?: AbortSignal): Promise<void> {
    if (this.closed) throw new FireweaveError('AlreadyClosed');
    if (this.client === undefined) {
      if (this.options.projectApiKey === undefined || this.options.projectApiKey.length === 0) {
        throw new FireweaveError('Configuration');
      }
      // Default-on host allowlist (H-1): the adapter is the egress point, so
      // it validates independently of the runtime-level config check.
      if (this.options.host !== undefined) {
        assertHostAllowed(this.options.host, this.options.allowedHosts);
      }
      // posthog-node is an optional peer dependency: load lazily so the main
      // entrypoint never depends on it.
      const mod = (await import('posthog-node')) as {
        PostHog: new (apiKey: string, options?: Record<string, unknown>) => unknown;
      };
      const clientOptions: Record<string, unknown> = {
        host: this.options.host,
        fetch: this.buildObservedFetch(),
        featureFlagsRequestMaxRetries: 0,
        fetchRetryCount: 0,
        flushAt: 100,
        disableCompression: true,
        // Belt-and-suspenders for non-snapshot APIs. Snapshot getFlag/isEnabled
        // still emit unless we read `_flags` (RB-2) — see fromSnapshot + gate.
        sendFeatureFlagEvent: false,
      };
      if (this.options.featureFlagsRequestTimeoutMs !== undefined) {
        clientOptions['featureFlagsRequestTimeoutMs'] = this.options.featureFlagsRequestTimeoutMs;
      }
      if (this.options.secretApiKey !== undefined) {
        clientOptions['personalApiKey'] = this.options.secretApiKey;
        clientOptions['enableLocalEvaluation'] = true;
        if (this.options.featureFlagsPollingInterval !== undefined) {
          clientOptions['featureFlagsPollingInterval'] = this.options.featureFlagsPollingInterval;
        }
      }
      const owned = new mod.PostHog(this.options.projectApiKey, clientOptions) as unknown as PostHogClientLike;
      // Drop vendor auto-exposures that lack Fireweave's arming marker. Explicit
      // recordExposure() sets fireweave.exposure=true so those still flow.
      this.client = this.gateVendorExposureCapture(owned);
    }
    if (this.options.waitForLocalDefinitions === true && this.client.waitForLocalEvaluationReady !== undefined) {
      const ready = await this.client.waitForLocalEvaluationReady(5000);
      if (!ready) throw new FireweaveError('NotReady');
    }
  }

  async resolve(flagKey: string, context: CanonicalContext, _options?: ResolveOptions): Promise<AdapterResolution> {
    if (this.closed) throw new FireweaveError('AlreadyClosed');
    if (this.client === undefined) throw new FireweaveError('NotReady');
    const distinctId = context.targetingKey ?? '';
    if (distinctId === '') {
      throw new FireweaveError('InvalidContext', {
        message: 'targeting key missing',
        openFeatureErrorCode: 'TARGETING_KEY_MISSING',
      });
    }

    const localOnly = this.options.onlyEvaluateLocally === true;
    if (localOnly && this.client.isLocalEvaluationReady !== undefined && !this.client.isLocalEvaluationReady()) {
      throw new FireweaveError('NotReady');
    }

    const personProperties: Record<string, string> = {};
    for (const [k, v] of Object.entries(context.attributes)) {
      // groups/groupProperties (plain aliases), fireweave.* carriers (ruling 13)
      // and $-prefixed system directives are not person properties.
      if (k === 'groups' || k === 'groupProperties' || k.startsWith('$') || k.startsWith('fireweave.')) {
        continue;
      }
      personProperties[k] = typeof v === 'string' ? v : JSON.stringify(v);
    }
    const groupProperties = this.extractGroupProperties(
      context.groupProperties ?? context.attributes['groupProperties'],
    );

    this.flagsObservations.delete(distinctId);
    const evalOptions: Parameters<PostHogClientLike['evaluateFlags']>[1] = {
      flagKeys: [flagKey],
      personProperties,
      disableGeoip: true,
    };
    if (context.groups !== undefined) evalOptions.groups = context.groups;
    if (groupProperties !== undefined) evalOptions.groupProperties = groupProperties;
    if (localOnly) evalOptions.onlyEvaluateLocally = true;

    const snapshot = await this.client.evaluateFlags(distinctId, evalOptions);
    const observation = this.flagsObservations.get(distinctId);

    if (!localOnly) {
      if (observation === undefined) {
        // RB-1: hybrid/local mode with a secret key (or definitions already
        // loaded) can satisfy evaluateFlags from the definitions poller without
        // hitting /flags. That is a successful local serve — not Network.
        if (this.canServeFromLocalSnapshot()) {
          return this.fromSnapshot(flagKey, snapshot);
        }
        // Remote-only (no secret / no definitions): missing observation means
        // the transport never reached /flags (offline).
        throw new FireweaveError('Network');
      }
      switch (observation.kind) {
        case 'abort':
          throw new FireweaveError('Timeout');
        case 'network':
          throw new FireweaveError('Network');
        case 'http':
          throw mapHttpStatus(observation.status ?? 0);
        case 'parse':
          throw new FireweaveError('MalformedResponse');
        case 'ok':
          return this.fromFlagsBody(flagKey, observation.body ?? {});
      }
    }

    // Local-only evaluation path: read from the snapshot without vendor emit.
    return this.fromSnapshot(flagKey, snapshot);
  }

  /** True when local definitions can satisfy a resolve without a /flags round-trip. */
  private canServeFromLocalSnapshot(): boolean {
    if (this.options.secretApiKey !== undefined) return true;
    if (this.definitionsLoaded) return true;
    if (this.client?.isLocalEvaluationReady?.() === true) return true;
    return false;
  }

  /**
   * Wrap an owned posthog-node client so vendor `$feature_flag_called` events
   * without Fireweave's arming marker are dropped (Go exposure-gate pattern).
   */
  private gateVendorExposureCapture(client: PostHogClientLike): PostHogClientLike {
    const rawCapture = client.capture.bind(client);
    client.capture = (props) => {
      if (
        props.event === '$feature_flag_called' &&
        props.properties?.['fireweave.exposure'] !== true
      ) {
        return;
      }
      rawCapture(props);
    };
    return client;
  }

  private fromFlagsBody(flagKey: string, body: FlagsV2Body): AdapterResolution {
    const quotaLimited = Array.isArray(body.quotaLimited) && body.quotaLimited.includes('feature_flags');
    const record = body.flags?.[flagKey];
    if (record === undefined) {
      const missing: AdapterResolution = { found: false };
      if (quotaLimited) missing.quotaLimited = true;
      return missing;
    }
    const variant = record.variant ?? undefined;
    const value: JsonValue = record.enabled ? (variant ?? true) : false;
    const resolution: AdapterResolution = {
      found: true,
      enabled: record.enabled,
      value,
    };
    if (variant !== undefined) resolution.variant = variant;
    if (record.reason?.code !== undefined) resolution.reasonCode = record.reason.code;
    if (record.reason?.condition_index !== undefined && record.reason.condition_index !== null) {
      resolution.conditionIndex = record.reason.condition_index;
    }
    if (record.metadata?.version !== undefined && record.metadata.version !== null) {
      resolution.version = record.metadata.version;
    }
    if (record.metadata?.id !== undefined && record.metadata.id !== null) {
      resolution.vendorFlagId = record.metadata.id;
    }
    if (record.metadata?.payload !== undefined && record.metadata.payload !== null) {
      resolution.payload = record.metadata.payload;
    }
    if (quotaLimited) resolution.quotaLimited = true;
    return resolution;
  }

  /**
   * Read a local/hybrid snapshot without calling getFlag/isEnabled (RB-2).
   * Those accessors unconditionally emit `$feature_flag_called` in posthog-node
   * and do not honor sendFeatureFlagEvent. Mirror Python: use `_flags` records.
   */
  private fromSnapshot(
    flagKey: string,
    snapshot: EvaluateFlagsSnapshot,
  ): AdapterResolution {
    const records = snapshot._flags;
    if (records === undefined) {
      // Structural fakes without `_flags` cannot be read side-effect-free.
      // Prefer missing over calling getFlag (which would emit vendor exposure).
      return { found: false };
    }
    const record = records[flagKey];
    if (record === undefined) return { found: false };
    const variant = record.variant ?? undefined;
    const variantStr = typeof variant === 'string' ? variant : undefined;
    const enabled = record.enabled;
    const resolution: AdapterResolution = {
      found: true,
      enabled,
      value: enabled ? (variantStr ?? true) : false,
    };
    if (variantStr !== undefined) resolution.variant = variantStr;
    if (record.reason !== undefined && record.reason !== null) {
      resolution.reasonCode = record.reason;
    }
    if (record.version !== undefined && record.version !== null) {
      resolution.version = record.version;
    }
    if (record.id !== undefined && record.id !== null) {
      resolution.vendorFlagId = record.id;
    }
    if (record.payload !== undefined && record.payload !== null) {
      resolution.payload = record.payload as JsonValue;
    } else {
      // getFlagPayload does not emit; safe enrichment when `_flags` omits payload.
      const payload = snapshot.getFlagPayload(flagKey);
      if (payload !== undefined && payload !== null) resolution.payload = payload as JsonValue;
    }
    if (this.definitionsStale) resolution.fromCache = true;
    return resolution;
  }

  private extractGroupProperties(
    value: Record<string, JsonValue> | JsonValue | undefined,
  ): Record<string, Record<string, string>> | undefined {
    if (value === undefined || value === null || typeof value !== 'object' || Array.isArray(value)) return undefined;
    const out: Record<string, Record<string, string>> = {};
    for (const [group, props] of Object.entries(value)) {
      if (props === null || typeof props !== 'object' || Array.isArray(props)) continue;
      const groupOut: Record<string, string> = {};
      for (const [k, v] of Object.entries(props)) {
        groupOut[k] = typeof v === 'string' ? v : JSON.stringify(v);
      }
      out[group] = groupOut;
    }
    return Object.keys(out).length > 0 ? out : undefined;
  }

  recordExposure(exposure: Exposure): void {
    if (this.client === undefined || this.closed) return;
    const properties: Record<string, unknown> = {
      $feature_flag: exposure.flagKey,
      $feature_flag_response: exposure.variant ?? exposure.value,
      'fireweave.exposure': true,
    };
    if (exposure.rolloutId !== undefined) properties['fireweave.rolloutId'] = exposure.rolloutId;
    if (exposure.changeId !== undefined) properties['fireweave.changeId'] = exposure.changeId;
    if (exposure.stampId !== undefined) properties['fireweave.stampId'] = exposure.stampId;
    this.client.capture({
      distinctId: exposure.targetingKey,
      event: '$feature_flag_called',
      properties,
    });
  }

  async flush(): Promise<void> {
    if (this.client === undefined || this.closed) return;
    try {
      await this.client.flush();
    } catch {
      // flush failures must not throw into shutdown paths
    }
  }

  recordSignal(signal: Signal): void {
    if (this.client === undefined || this.closed) return;
    const properties: Record<string, unknown> = {
      'fireweave.signal': true,
      kind: signal.kind,
      name: signal.name,
    };
    if (signal.status !== undefined) properties['status'] = signal.status;
    if (signal.errorKind !== undefined) properties['errorKind'] = signal.errorKind;
    if (signal.message !== undefined) properties['message'] = signal.message;
    if (signal.value !== undefined) properties['value'] = signal.value;
    if (signal.rolloutId !== undefined) properties['fireweave.rolloutId'] = signal.rolloutId;
    if (signal.changeId !== undefined) properties['fireweave.changeId'] = signal.changeId;
    if (signal.stampId !== undefined) properties['fireweave.stampId'] = signal.stampId;
    if (signal.flagKey !== undefined) properties['flagKey'] = signal.flagKey;
    if (signal.variant !== undefined) properties['variant'] = signal.variant;
    this.client.capture({
      distinctId: signal.targetingKey ?? signal.rolloutId ?? 'fireweave-sdk',
      event: `$fw_signal_${signal.kind}`,
      properties,
    });
  }

  async shutdown(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    if (this.client !== undefined && this.ownsClient) {
      try {
        // Configured deadline (M-1/L-4): no more hardcoded 2 s — the vendor
        // client gets the same bound the config/capabilities advertise.
        await this.client.shutdown(this.options.shutdownTimeoutMs ?? DEFAULT_ADAPTER_SHUTDOWN_TIMEOUT_MS);
      } catch {
        // never throw from shutdown
      }
    }
  }

  isDefinitionsStale(): boolean {
    return this.definitionsStale;
  }

  features(): AdapterRuntimeFeatures {
    const local = this.options.secretApiKey !== undefined;
    return {
      remoteEvaluation: !(this.options.onlyEvaluateLocally === true),
      localEvaluation: local,
      localOnly: this.options.onlyEvaluateLocally === true,
      exposureEmission: true,
      sideEffectFreeReads: true,
      groupAnalytics: true,
    };
  }
}
